superset/connectors/base/models.py

```diff
@@ -213,6 +230,87 @@ def data(self) -> Dict[str, Any]:
             "select_star": self.select_star,
         }
 
+    @staticmethod
+    def get_metric_name_from_metric(
+        metric: Union[str, Dict[str, Any]]
+    ) -> Optional[str]:
+        if isinstance(metric, str):
+            return metric
+
+        if isinstance(metric, dict) and metric.get("metric_name"):
+            return metric.get("metric_name")
+
+        return None
```

Comment:
I might be wrong here, but isn't this function already implemented in `from superset.utils.core import get_metric_name()`?

Reply:
I don't actually think so, that function gets the "label" field off a metric instead of the "metric_name" one.  And I don't see the `label` param anywhere in the `BaseMetric` class, only the `metric_name` one

Reply:
I looked further, and it seems like the metric field is a string if it's a `metric_name`, but an object with a label param if it's an `adhoc_metric`. Although I'm not sure if these labels would actually show up in the datasource, I'll add logic to handle them

Reply:
updated!

Reply:
The weird thing is what you're describing sounds more or less exactly like the other metrics that `get_metric_name` deals with. We don't need to deal with that now as it's outside the scope of this PR, but we definitely need to figure this out in the long run.

Reply:
I think it's probably a confusion between "metrics" in form data and the "metric" model in Superset's backend. I think the "metrics" in form data are either `metric.metric_name` where `metric` is the metric model or an `AdHocMetric` in the JS frontend, which might not have any association with the metric model (i'm uncertain here). 

We could probably improve this by standardizing on a single typedef/naming scheme and then migrating all the chart form_data, but this isn't an easy thing to do (and will only get harder once charts are more pluggable and we can't guarantee what type of data might be in form_data)

Reply:
I've been digesting this today, and I still think something isn't right here. If you look at `viz.py`, it's full of references to metrics in form_data. Copy/paste from `viz.py`: `utils.get_metric_names(self.form_data.get("metrics") or [])`. Are these really regular ad-hoc metrics that have been generated by Superset controls? I put a break point on `viz.py` with a viz with an ad-hoc metric, and that had the metric name in `label`, not `metric_name`. Is there a chance that we have different metric signature conventions in different parts of the codebase?